### PR TITLE
feat: Integrate flash attention

### PR DESCRIPTION
**Pull request by OrangeSodahub**
**Goal**: integrate the flash attention (official [repo](https://github.com/HazyResearch/flash-attention)) into the CLIP Model (will replace the `torch.nn.MultiheadAttention`).

**Required**: CUDA11 and PyTorch>=1.8 and Turing or Ampere GPUs (e.g., A100, RTX 3090, T4, RTX 2080)

**Performance**:

Test model (on `RTX3080`): `ViT-L-14::laion2b-s32b-b82k`

Times = **100**, using causal mask

| shape             | baseline(s) | flash_attn(s) | speed up (x) | mean diff |
|-------------------|-------------|---------------|--------------|-----------|
| (1, 77)           | 0.66602     | 0.55968       | 1.1899       | 0.000599  |
| (2, 77)           | 0.69619     | 0.60348       | 1.1536       | 0.000570  |
| (4, 77)           | 0.75353     | 0.65664       | 1.1225       | 0.000573  |
| (8, 77)           | 0.73511     | 0.64203       | 1.1449       | 0.000530  |
| (16, 77)          | 0.78381     | 0.72705       | 1.0780       | 0.000545  |
| (1, 3, 224, 224)  | 1.25425     | 1.10564       | 1.1344       | 0.001713  |
| (2, 3, 224, 224)  | 1.32202     | 1.22698       | 1.0775       | 0.001739  |
| (4, 3, 224, 224)  | 2.36880     | 2.11119       | 1.0926       | 0.001227  |
| (8, 3, 224, 224)  | 3.88597     | 3.76666       | 1.0316       | 0.001441  |
| (16, 3, 224, 224) | 7.84954     | 7.20817       | 1.0889       | 0.001533  |

Test model (on `Tesla T4`): `ViT-L-14::laion2b-s32b-b82k`

Times = **100**, using causal mask

|       shape       | baseline(s) | flash_attn(s) | speed up (x) | mean diff |
|:-----------------:|:-----------:|:-------------:|:------------:|:---------:|
| (1, 77)           |     0.81193 |       0.66437 |      1.22211 |   0.00066 |
| (2, 77)           |      0.8264 |       0.70035 |      1.17998 |    0.0007 |
| (4, 77)           |     0.81998 |       0.69887 |       1.1733 |   0.00064 |
| (8, 77)           |     1.05975 |       0.85742 |      1.23597 |   0.00054 |
| (16, 77)          |     2.12992 |       1.68367 |      1.26504 |   0.00057 |
| (1, 3, 224, 224)  |     2.00593 |       1.34507 |      1.49131 |   0.00155 |
| (2, 3, 224, 224)  |     3.74493 |       2.29818 |      1.62952 |   0.00122 |
| (4, 3, 224, 224)  |     7.35365 |       4.44447 |      1.65456 |   0.00159 |
| (8, 3, 224, 224)  |    14.63006 |       9.05604 |       1.6155 |   0.00135 |
| (16, 3, 224, 224) |    29.63732 |      18.29142 |      1.62028 |   0.00155 |

Test model (on `Tesla T4`): `ViT-B-32::laion2b-s34b-b79k`

Times = **100**, using causal mask

|       shape       | baseline(s) | flash_attn(s) | speed up (x) | mean diff |
|:-----------------:|:-----------:|:-------------:|:------------:|:---------:|
| (1, 77)           |     0.42692 |       0.37867 |       1.1274 |   0.00076 |
| (2, 77)           |     0.46548 |       0.42958 |      1.08358 |   0.00077 |
| (4, 77)           |     0.49818 |       0.46378 |      1.07417 |   0.00079 |
| (8, 77)           |     0.48738 |       0.45324 |       1.0753 |   0.00068 |
| (16, 77)          |      0.4764 |       0.44315 |      1.07502 |   0.00067 |
| (1, 3, 224, 224)  |      0.4349 |       0.40392 |       1.0767 |   0.00054 |
| (2, 3, 224, 224)  |      0.4445 |       0.42513 |      1.04555 |   0.00058 |
| (4, 3, 224, 224)  |     0.43605 |       0.41524 |      1.05009 |   0.00054 |
| (8, 3, 224, 224)  |     0.47367 |       0.45316 |      1.04527 |   0.00064 |
| (16, 3, 224, 224) |     0.51586 |       0.50555 |       1.0204 |   0.00058 |

Test model (on `Tesla T4`): `ViT-B-16::laion400m_e31`

Times = **100**, using causal mask

|       shape       | baseline(s) | flash_attn(s) | speed up (x) | mean diff |
|:-----------------:|:-----------:|:-------------:|:------------:|:---------:|
| (1, 77)           |     0.42736 |       0.38143 |      1.12042 |   0.00051 |
| (2, 77)           |     0.46499 |         0.432 |      1.07635 |   0.00025 |
| (4, 77)           |     0.50052 |       0.46798 |      1.06953 |   0.00045 |
| (8, 77)           |     0.48717 |       0.45675 |      1.06659 |   0.00043 |
| (16, 77)          |     0.47855 |       0.44717 |      1.07017 |    0.0005 |
| (1, 3, 224, 224)  |     0.45754 |       0.38839 |      1.17803 |   0.00044 |
| (2, 3, 224, 224)  |     0.50152 |       0.44523 |      1.12641 |   0.00042 |
| (4, 3, 224, 224)  |     0.56444 |       0.51215 |       1.1021 |   0.00044 |
| (8, 3, 224, 224)  |     0.93341 |       0.82501 |      1.13139 |   0.00045 |
| (16, 3, 224, 224) |     1.77167 |       1.58981 |      1.11439 |   0.00044 |

Test model (on `Tesla T4`): `ViT-g-14::laion2b-s12b-b42k`

Times = **100**, using causal mask

|       shape       | baseline(s) | flash_attn(s) | speed up (x) | mean diff |
|:-----------------:|:-----------:|:-------------:|:------------:|:---------:|
| (1, 77)           |      0.8894 |        0.7984 |      1.11397 |   0.00093 |
| (2, 77)           |     0.89632 |       0.83137 |      1.07812 |   0.00079 |
| (4, 77)           |     0.93494 |       0.87282 |      1.07117 |    0.0008 |
| (8, 77)           |     1.15918 |       1.08828 |      1.06514 |   0.00073 |
| (16, 77)          |     2.05242 |       1.92306 |      1.06727 |   0.00058 |
| (1, 3, 224, 224)  |     1.94179 |       1.85599 |      1.04623 |   0.00115 |
| (2, 3, 224, 224)  |     3.29228 |       3.10451 |      1.06048 |   0.00109 |
| (4, 3, 224, 224)  |     5.87704 |       5.55928 |      1.05715 |   0.00115 |
| (8, 3, 224, 224)  |    10.37121 |        9.8581 |      1.05204 |   0.00101 |
| (16, 3, 224, 224) |    20.85575 |      19.86552 |      1.04984 |   0.00112 |

Test model (on `Tesla T4`): `ViT-H-14::laion2b-s32b-b79k`

Times = **100**, using causal mask

|       shape       | baseline(s) | flash_attn(s) | speed up (x) | mean diff |
|:-----------------:|:-----------:|:-------------:|:------------:|:---------:|
| (1, 77)           |     0.90747 |       0.81617 |      1.11186 |    0.0008 |
| (2, 77)           |     0.90999 |       0.84415 |      1.07799 |   0.00077 |
| (4, 77)           |     0.97304 |        0.9068 |      1.07305 |   0.00053 |
| (8, 77)           |     2.55404 |       2.48595 |      1.02738 |   0.00067 |
| (16, 77)          |     4.36827 |       4.11261 |      1.06216 |   0.00067 |
| (1, 3, 224, 224)  |     1.33663 |       1.26552 |      1.05619 |    0.0009 |
| (2, 3, 224, 224)  |       2.273 |       2.12032 |        1.072 |    0.0008 |
| (4, 3, 224, 224)  |     4.04465 |       3.78626 |      1.06824 |   0.00082 |
| (8, 3, 224, 224)  |     7.23102 |       6.80665 |      1.06234 |    0.0008 |
| (16, 3, 224, 224) |    13.64971 |       12.8397 |      1.06308 |   0.00087 |

### CODE REVIEW

File: server/clip_server/model/model.py [hunk 1]
```diff
@@ -277,11 +278,17 @@ def __init__(
         scale_heads: bool = False,
         scale_attn: bool = False,
         scale_fc: bool = False,
+        use_flash: bool = True,
```
Comment (numb3r3) [inline server/clip_server/model/model.py:281]: I would like to take `flash_attention` as the default behavior by detecting the depends.

Comment (numb3r3) [reply server/clip_server/model/model.py:281]: So, can we have a helper function to detect whether flash_attention is usable?

File: server/clip_server/model/flash_attention.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
+import torch
```
Comment (numb3r3) [inline server/clip_server/model/flash_attention.py:1]: Is this module code implemented by us? If not, please provide the copyright info.

**Review by numb3r3 — CHANGES_REQUESTED** [review]
Please also present the performance and memory difference with flash_attention in the description.

**Comment by numb3r3** [comment]
And what's more, please confirm that the flash_attention can load the pre-trained parameters, and result the consistency output with `nn.MultiHeadAttention`

File: server/clip_server/model/flash_attention.py (hunk 2 above)
Comment (OrangeSodahub) [reply server/clip_server/model/flash_attention.py:1]: By us

**Comment by OrangeSodahub** [comment]
> And what's more, please confirm that the flash_attention can load the pre-trained parameters, and result the consistency output with `nn.MultiHeadAttention`



> And what's more, please confirm that the flash_attention can load the pre-trained parameters, and result the consistency output with `nn.MultiHeadAttention`



> And what's more, please confirm that the flash_attention can load the pre-trained parameters, and result the consistency output with `nn.MultiHeadAttention`

I've confirmed the results' correctness in this [repo](https://github.com/OrangeSodahub/Openclip-flashattn). There exists only errors of 1E-3 level.

**Comment by numb3r3** [comment]
@OrangeSodahub Please also present the performance under full-precision inference in the `Openclip-flashattn` repo

**Comment by OrangeSodahub** [comment]
> @OrangeSodahub Please also present the performance under full-precision inference in the `Openclip-flashattn` repo

Flash attention does not support fp32 see: https://github.com/HazyResearch/flash-attention#alpha-release-01

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -158,6 +158,7 @@ jobs:
         run: |
           python -m pip install --upgrade pip
           python -m pip install wheel pytest pytest-cov nvidia-pyindex
+          python -m pip install git+https://github.com/HazyResearch/flash-attention.git
```
Comment (numb3r3) [inline .github/workflows/ci.yml:161]: please also add this package in `setup.py`

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 3 above)
Comment (OrangeSodahub) [reply .github/workflows/ci.yml:161]: That's a problem. `setup.py` cannot install packages via git. `flash attn` has not publish pip yet.

Comment (numb3r3) [reply .github/workflows/ci.yml:161]: Then, we should also check whether `flash-attention` is installed when we want to use it, like:
```
try:
    import flash_attn
except:
    FLASH_ATTENTION_AVAILABLE = False
```

**Review by numb3r3 — CHANGES_REQUESTED** [review]

**Comment by OrangeSodahub** [comment]
Now the situation is: the gpu device only has the CUDA 11.0 and only pytorch <=1.7.1 matches. But <1.8 does not support `flash_attn`.

File: server/clip_server/model/model.py [hunk 4]
```diff
@@ -26,6 +26,14 @@
 from open_clip.utils import freeze_batch_norm_2d
 from open_clip.factory import _MODEL_CONFIGS
 
+# Use flash attention
+try:
+    from clip_server.model.flash_attention import MultiheadAttention
```
Comment (numb3r3) [inline server/clip_server/model/model.py:31]: We also need to check the compatibility of cuda and pytorch here.

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: .github/workflows/cuda/cu113-Linux-env.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+CUDA_HOME=/usr/local/cuda-11.3
+LD_LIBRARY_PATH=${CUDA_HOME}/lib64:${LD_LIBRARY_PATH}
+PATH=${CUDA_HOME}/bin:${PATH}
+
+export FORCE_CUDA=1
+export TORCH_CUDA_ARCH_LIST="3.5;5.0+PTX;6.0;7.0;7.5;8.0;8.6"
+export CUDA_HOME=/usr/local/cuda-11.3
```
Comment (numb3r3) [inline .github/workflows/cuda/cu113-Linux-env.sh:9]: Reject this change, this is overkill. Just add validation at the beginning of the `model.py`.

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: server/clip_server/model/model.py (hunk 4 above)
Comment (numb3r3) [reply server/clip_server/model/model.py:31]: Insert the cuda and pytorch version checking here.

**Review by numb3r3 — CHANGES_REQUESTED** [review]

File: .github/workflows/cuda/cu113-Linux-env.sh (hunk 5 above)
Comment (OrangeSodahub) [reply .github/workflows/cuda/cu113-Linux-env.sh:9]: It's just a test. see the commit `fix: passby flash attention`. And I think no need to check this. Once import succeeds, they should match.

Comment (OrangeSodahub) [reply .github/workflows/cuda/cu113-Linux-env.sh]: https://github.com/jina-ai/clip-as-service/actions/runs/3459017298/jobs/5774137064
